docs/decisions/0024-planner-telemetry-enhancement.md

```diff
@@ -0,0 +1,366 @@
+    - Dimensions: ComponentType, ComponentName, Service ID, Model ID
+    - Example:
+      | ComponentType | ComponentName | Value |
+      |---|---|---|
+      | Function | WritePoem | 1m
+      | Function | TellJoke | 1m
+      | Function | WriteAndTellJoke | 1.5m
+      | Planner | CreateHandlebarsPlan | 2m
+  - Success count for planners
+    - Description: A planner run is considered successful when it generates a valid plan.
+    - Dimensions: ComponentType, ComponentName, Service ID, Model ID
+    - Example:
+      | ComponentType | ComponentName | Fail | Success
+      |---|---|---|---|
+      | Planner | CreateHandlebarsPlan | 5 | 95
+      | Planner | CreateHSequentialPlan | 20 | 80
+  - Success count for plans
+    - Description: A plan execution is considered successful when all steps in the plan are executed successfully.
+    - Dimensions: ComponentType, ComponentName, Service ID, Model ID
+    - Example:
+      | ComponentType | ComponentName | Fail | Success
+      |---|---|---|---|
+      | Plan | HandlebarsPlan | 5 | 95
+      | Plan | SequentialPlan | 20 | 80
+
+## Considered Options
+
+- Function hooks
+  - Inject logic to functions that will get executed before or after a function is invoked.
+- Instrumentation
+  - Logging
+  - Metrics
+  - Traces
+
+## Other Considerations
+
+SK currently tracks token usage metrics in connectors; however, these metrics are not categorized. Consequently, developers cannot determine token usage for different operations. To address this issue, we propose the following two approaches:
+
+- Bottom-up: Propagate token usage information from connectors back to the functions, along with model results.
+- Top-down: Propagate function information down to the connectors, enabling them to tag metric items with function information.
+
+We have decided to implement the bottom-up approach for the following reasons:
+
+1. SK is already configured to propagate token usage information from connectors via `ModelResult`. We simply need to extend the list of items that get propagated, such as model information.
+2. Currently, SK does not have a method for passing function information down to the connector level. Although we considered using [baggage](https://opentelemetry.io/docs/concepts/signals/baggage/#:~:text=In%20OpenTelemetry%2C%20Baggage%20is%20contextual%20information%20that%E2%80%99s%20passed,available%20to%20any%20span%20created%20within%20that%20trace.) as a means of propagating information downward, experts from the OpenTelemetry team advised against this approach due to security concerns.
+
+With the bottom-up approach, we need to add model information to `IResultBase`:
+
+```csharp
+/// <summary>
+/// Interface for model results
+/// </summary>
+public interface IResultBase
+{
+  /// <summary>
+  /// Model name or Id
+  /// </summary>
+  string ModelId { get; }
```

Comment:
Every time we add a new member to an interface, we break all its implementations, as they must implement it. For telemetry purposes, we need a more extensible solution for passing information up the stack. Today, we need to track the number of tokens, and tomorrow we may need to track the prompt-response relevance score.

Reply:
Yes, I `Dictionary` would be a more flexible approach and we can add extension methods for well known entries

Reply:
Updated the ADR with the option to carry the model id in `ModelResult`.